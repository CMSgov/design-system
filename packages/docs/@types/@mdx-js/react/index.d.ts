declare module '@mdx-js/react' {
  const MDXProvider: React.ComponentType;
  export { MDXProvider };
}
