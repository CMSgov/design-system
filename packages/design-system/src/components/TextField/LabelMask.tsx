import { Children, cloneElement } from 'react';
import type * as React from 'react';
import { useLabelMask, MaskFunction } from './useLabelMask';

export interface LabelMaskProps {
  /**
   * Must contain a `TextField` component
   */
  children: React.ReactNode;
  /**
   * Providing a function here will turn the text field into a label-masked field, where
   * the user input is formatted in a label as the user types and then the input field
   * itself is automatically formatted when the user shifts focus away from the input.
   * A custom function can be given, or one of the following built-in functions can be
   * imported from the design system and passed to this component: PHONE_MASK, ZIP_MASK,
   * SSN_MASK, and CURRENCY_MASK. See
   * [Label-masked field](https://design.cms.gov/components/text-field/label-masked-field/)
   * documentation page for more information.
   */
  labelMask?: MaskFunction;
}

const LabelMask = (props: LabelMaskProps) => {
  const field = Children.only(props.children as React.ReactElement);
  const { labelMask, inputProps } = useLabelMask(props.labelMask, field.props);
  const input = cloneElement(field, inputProps);

  return (
    <>
      {labelMask}
      {input}
    </>
  );
};

export default LabelMask;
