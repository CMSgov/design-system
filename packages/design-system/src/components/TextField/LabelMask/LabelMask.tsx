import React from 'react';

function maskDate(rawInput = '', valueOnly = false): any {
  const RE_DATE = /^(\d{1,2})[\D]?(\d{1,2})?[\D]?(\d{1,4})?/;
  const match = RE_DATE.exec(rawInput);
  let formattedDate = '';
  if (match) {
    const [month, day, year] = match.slice(1);
    formattedDate = [
      // We treat all non-numeric characters as a delimiter. If they're using a
      // delimiter after a month or day, we interpret that as the user supplying
      // a single digit for month or day, which we will automatically pad for them.
      month && month.padStart(2, '0'),
      day && day.padStart(2, '0'),
      year,
    ]
      .filter((s) => s)
      .join('/');
  }
  if (valueOnly) {
    return formattedDate;
  }
  const hint = 'MM/DD/YYYY';
  const hintSub = hint.substring(formattedDate.length);
  return formattedDate + hintSub;
}

export const BuiltInMask = {
  DATE: maskDate,
};

export interface LabelMaskProps {
  /**
   * Must contain a `TextField` component
   */
  children: React.ReactNode;
  labelMask?: (rawInput: string, valueOnly?: boolean) => string;
}

const LabelMask = ({ children, labelMask }: LabelMaskProps) => {
  if (typeof labelMask === 'string') {
    if (labelMask === BuiltInMask.DATE) {
      labelMask = maskDate;
    }
  }

  /**
   * Get the child text field. Called as a method so that
   * updates to the field cause the mask to re-render
   * @returns {React.ReactElement} Child TextField
   */
  const field = (): React.ReactElement => React.Children.only(children as React.ReactElement);
  const value = field().props.value;

  /**
   * @param {Object} evt
   * @param {React.Element} field - Child TextField
   */
  const handleChange = (evt: React.ChangeEvent<HTMLInputElement>) => {
    if (typeof field().props.onChange === 'function') {
      return field().props.onChange(evt);
    }
  };

  const modifiedTextField = React.cloneElement(field(), {
    defaultValue: undefined,
    onChange: (e) => handleChange(e),
    onBlur: (e) => {
      const maskedValue = labelMask(value, true);
      e.currentTarget.value = maskedValue;
      e.target.value = maskedValue;
      handleChange(e);

      if (typeof field().props.onBlur === 'function') {
        return field().props.onBlur(e);
      }
    },
    type: 'text',
    inputMode: 'numeric',
  });

  return (
    <div className="ds-c-label-mask">
      {modifiedTextField}
      <div className="ds-c-label-mask__mask" aria-hidden="true">
        <span className="ds-c-label-mask__mask--active">{labelMask(value)}</span>
        <span className="ds-c-label-mask__mask--inactive">{labelMask('')}</span>
      </div>
    </div>
  );
};

export default LabelMask;
