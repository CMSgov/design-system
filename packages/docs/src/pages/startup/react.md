---
title: Using React components
weight: 3
---

The examples below assume you've installed the design system using NPM and have already setup your build system.

<h4 class="ds-h4 ds-u-font-size--base">Default imports</h4>

Individual components can be imported from their individual export file.

```jsx
import Button from '@cmsgov/design-system-core/dist/components/Button/Button';
```

<h4 class="ds-h4 ds-u-font-size--base">Named imports</h4>

Components can also be imported using the shorter syntax below.

<div class="ds-c-alert ds-c-alert--warn ds-u-margin-bottom--2 ds-u-font-size--small">
  <div class="ds-c-alert__body">
    <h3 class="ds-c-alert__heading ds-u-font-size--base">Performance note</h3>
    <p class="ds-c-alert__text">
      This approach may result in a much larger file than you intend. Depending on what module bundler you use, all of the design system's React components may be included in the bundled file even if you didn't specifically import them. This can be avoided by enabling features like [tree shaking in Webpack](https://webpack.js.org/guides/tree-shaking/).
    </p>
  </div>
</div>

```jsx
import { Button } from '@cmsgov/design-system-core';
```

[View an example](https://github.com/CMSgov/design-system/blob/master/examples/react-app/src/scripts/index.js)

<h2 id="examples" class="ds-h2 ds-u-color--primary-darker">Examples</h2>

Additional examples of the design system in use can be viewed on GitHub. These projects demonstrate the various ways you can incorporate the design system into your development process and various use cases.

<a href="https://github.com/CMSgov/design-system/tree/master/examples/" class="ds-c-button">Browse example projects</a>

<h2 id="need-help" class="ds-h2 ds-u-color--primary-darker">Need help or ran into an issue?</h2>

If you're having trouble installing or setting up the design system, or if you think you've found a bug, feel free to [open an issue on GitHub](https://github.com/CMSgov/design-system/issues).
