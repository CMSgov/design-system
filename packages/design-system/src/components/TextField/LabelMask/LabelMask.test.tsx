import React, { useState } from 'react';
import LabelMask, { BuiltInMask } from './LabelMask';
import { render, fireEvent } from '@testing-library/react';

const TestLabelMask = () => {
  const [value, setValue] = useState('');
  return (
    <LabelMask labelMask={BuiltInMask.DATE}>
      <input type="text" value={value} onChange={(e) => setValue(e.currentTarget.value)} />
    </LabelMask>
  );
};

describe('Label mask', function () {
  it('renders default date mask, MM/DD/YYYY, when no input value set', () => {
    const { container, asFragment } = render(<TestLabelMask />);
    const mask = container.querySelector('.ds-c-label-mask__mask--active');
    const input = container.querySelector('input');

    expect(input.value).toBe('');
    expect(mask.textContent).toBe('MM/DD/YYYY');

    expect(asFragment()).toMatchSnapshot();
  });
  describe('updates label mask to reflect', () => {
    it('complete input value set', () => {
      const { container, asFragment } = render(<TestLabelMask />);
      const data = '12345678';
      const maskText = '12/34/5678';

      const mask = container.querySelector('.ds-c-label-mask__mask--active');
      const input = container.querySelector('input');

      fireEvent.change(input, { target: { value: data } });

      expect(input).toHaveValue(data);
      expect(mask.textContent).toBe(maskText);

      expect(asFragment()).toMatchSnapshot();
    });
    it('partial input value set', () => {
      const { container, asFragment } = render(<TestLabelMask />);
      const data = '1234';

      const mask = container.querySelector('.ds-c-label-mask__mask--active');
      const input = container.querySelector('input');

      fireEvent.change(input, { target: { value: data } });

      expect(input).toHaveValue(data);
      expect(mask.textContent).toBe('12/34/YYYY');

      expect(asFragment()).toMatchSnapshot();
    });
    it('padded values when incomplete input value set', () => {
      const { container, asFragment } = render(<TestLabelMask />);
      const data = '1';
      const maskText = '01/DD/YYYY';

      const mask = container.querySelector('.ds-c-label-mask__mask--active');
      const input = container.querySelector('input');

      fireEvent.change(input, { target: { value: data } });

      expect(input).toHaveValue(data);
      expect(mask.textContent).toBe(maskText);

      expect(asFragment()).toMatchSnapshot();
    });
  });
  it('formats input value onBlur', () => {
    const { container, asFragment } = render(<TestLabelMask />);
    const data = '12345678';
    const formattedData = '12/34/5678';

    const input = container.querySelector('input');

    fireEvent.change(input, { target: { value: data } });
    fireEvent.blur(input);

    expect(input).toHaveValue(formattedData);

    expect(asFragment()).toMatchSnapshot();
  });
});
