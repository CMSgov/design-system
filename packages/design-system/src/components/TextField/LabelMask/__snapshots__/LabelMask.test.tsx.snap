// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label mask formats input value onBlur 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
  >
    <input
      inputmode="numeric"
      type="text"
      value="12345678"
    />
    <div
      aria-hidden="true"
      class="ds-c-label-mask__mask"
    >
      <span
        class="ds-c-label-mask__mask--active"
      >
        12/34/5678
      </span>
      <span
        class="ds-c-label-mask__mask--inactive"
      >
        MM/DD/YYYY
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Label mask renders default date mask, MM/DD/YYYY, when no input value set 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
  >
    <input
      inputmode="numeric"
      type="text"
      value=""
    />
    <div
      aria-hidden="true"
      class="ds-c-label-mask__mask"
    >
      <span
        class="ds-c-label-mask__mask--active"
      >
        MM/DD/YYYY
      </span>
      <span
        class="ds-c-label-mask__mask--inactive"
      >
        MM/DD/YYYY
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Label mask updates date mask to reflect complete input value set 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
  >
    <input
      inputmode="numeric"
      type="text"
      value="12345678"
    />
    <div
      aria-hidden="true"
      class="ds-c-label-mask__mask"
    >
      <span
        class="ds-c-label-mask__mask--active"
      >
        12/34/5678
      </span>
      <span
        class="ds-c-label-mask__mask--inactive"
      >
        MM/DD/YYYY
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Label mask updates date mask to reflect input value set 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
  >
    <input
      inputmode="numeric"
      type="text"
      value="12345678"
    />
    <div
      aria-hidden="true"
      class="ds-c-label-mask__mask"
    >
      <span
        class="ds-c-label-mask__mask--active"
      >
        12/34/5678
      </span>
      <span
        class="ds-c-label-mask__mask--inactive"
      >
        MM/DD/YYYY
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Label mask updates date mask to reflect partial input value set 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
  >
    <input
      inputmode="numeric"
      type="text"
      value="1234"
    />
    <div
      aria-hidden="true"
      class="ds-c-label-mask__mask"
    >
      <span
        class="ds-c-label-mask__mask--active"
      >
        12/34/YYYY
      </span>
      <span
        class="ds-c-label-mask__mask--inactive"
      >
        MM/DD/YYYY
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Label mask updates label mask to reflect complete input value set 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
  >
    <input
      inputmode="numeric"
      type="text"
      value="12345678"
    />
    <div
      aria-hidden="true"
      class="ds-c-label-mask__mask"
    >
      <span
        class="ds-c-label-mask__mask--active"
      >
        12/34/5678
      </span>
      <span
        class="ds-c-label-mask__mask--inactive"
      >
        MM/DD/YYYY
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Label mask updates label mask to reflect padded values when incomplete input value set 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
  >
    <input
      inputmode="numeric"
      type="text"
      value="1"
    />
    <div
      aria-hidden="true"
      class="ds-c-label-mask__mask"
    >
      <span
        class="ds-c-label-mask__mask--active"
      >
        01/DD/YYYY
      </span>
      <span
        class="ds-c-label-mask__mask--inactive"
      >
        MM/DD/YYYY
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Label mask updates label mask to reflect partial input value set 1`] = `
<DocumentFragment>
  <div
    class="ds-c-label-mask"
  >
    <input
      inputmode="numeric"
      type="text"
      value="1234"
    />
    <div
      aria-hidden="true"
      class="ds-c-label-mask__mask"
    >
      <span
        class="ds-c-label-mask__mask--active"
      >
        12/34/YYYY
      </span>
      <span
        class="ds-c-label-mask__mask--inactive"
      >
        MM/DD/YYYY
      </span>
    </div>
  </div>
</DocumentFragment>
`;
